### ISSUE TITLE
JSON retrieved through KernelClient has mixed single quotes (and sometimes mixed single/double)

### ISSUE THREAD

**Issue by tcpie** [jupyter/jupyter_client#187]
I am running into an issue where the JSON I get back contains both single and double quotes as delimiters. (this is with the default python3 IPython (Jupyter?) kernel). Use of single quotes as delimiters [is invalid JSON](http://www.json.org/) (!). My application is written in C++ (with Qt) and the standard Qt JSON parser has issues with that. In some cases I can get away with replacing all single quotes with double ones, but not always.

As an example, see the `execute_input` message I receive after running `print('Hello, World!')`:

> {'msg_id': 'dcef6f6d-5f1c-4476-a226-ed60eba40722', 'buffers': [], 'parent_header': {'msg_id': '24accdff-0f00-4b10-ba35-1bd2983ec4c7', 'session': 'd42bd8d2-850e-4bae-8349-45ddbfea4705', 'version': '5.0', 'date': datetime.datetime(2016, 8, 21, 7, 33, 36, 294979), 'msg_type': 'execute_request', 'username': 'tcpie'}, 'msg_type': 'execute_input', 'content': {'code': "print('Hello, World!')", 'execution_count': 1}, 'metadata': {}, 'header': {'msg_id': 'dcef6f6d-5f1c-4476-a226-ed60eba40722', 'session': '0c3ffbcb-dd20-4804-bc28-46110bb76141', 'version': '5.0', 'date': datetime.datetime(2016, 8, 21, 7, 33, 36, 297474), 'msg_type': 'execute_input', 'username': 'tcpie'}}

Please have the JSON use **double** quotes, as per the specification, and properly escape double quotes within data.

**Comment by takluyver** [comment]
Our JSON serialisation uses the Python standard library, and I'm pretty sure it's entirely JSON compliant. What you're seeing is the Python `repr()` of an object loaded *from* JSON. While this looks superficially similar, it's not meant to be valid JSON.

If you're using `jupyter_client`, it handles parsing the JSON for you, so there should be no need to use the Qt JSON parser. Can you show the code that you're running?

**Comment by tcpie** [comment]
Hey takluyver,

Thank you very much for your quick reply, I did not know that. 

I retrieve my (iopub) messages as follows: (C++)
```
QString ret = ""
PyObject *args = Py_BuildValue("()");
PyObject *keywords = PyDict_New();
PyObject* method = PyObject_GetAttrString(this->kernel_client, "get_iopub_msg");
PyObject* msg = PyEval_CallObjectWithKeywords(method, args, keywords);

if (msg == NULL)
    return ret;

PyObject* msg_str = PyObject_Str(msg);
ret = QString(PyUnicode_AsUTF8(msg_str));
```
And then I parse the `ret` object with the Qt JSON parser.

I did not know `jupyter_client` parses the messages already. The documentation says: 

> get_iopub_msg(*args, **kwargs)
>     Get a message from the iopub channel

It does not say the return type of the function. That's why I (wrongly) assumed it was a string.

**Comment by takluyver** [comment]
Yeah, `msg` is a Python dictionary, so when you call `PyObject_Str`, you're getting the Python representation of it. Most of the APIs work with deserialised messages, because handling the deserialisation correctly is part of jupyter_client's responsibility (the wire format actually has four separate pieces of JSON, which it parses and reassembles into one message). We do need to improve the high-level docs on using jupyter_client, though.
